api/provider/aws/aws_test/load_balancer_create_test.go

```diff
@@ -0,0 +1,335 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLoadBalancerCreate(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+	mockConfig.EXPECT().VPC().Return("vpc_id").AnyTimes()
+	mockConfig.EXPECT().Region().Return("region").AnyTimes()
+	mockConfig.EXPECT().AccountID().Return("123456789012")
+	mockConfig.EXPECT().PrivateSubnets().Return([]string{"priv1", "priv2"}).AnyTimes()
+
+	defer provider.SetEntityIDGenerator("lb_id")()
+
+	req := models.CreateLoadBalancerRequest{
+		LoadBalancerName: "lb_name",
+		EnvironmentID:    "env_id",
+		IsPublic:         false,
+		Ports: []models.Port{
+			{
+				CertificateName: "cert",
+				ContainerPort:   80,
+				HostPort:        80,
+				Protocol:        "http",
+			},
+			{
+				CertificateName: "cert",
+				ContainerPort:   443,
+				HostPort:        443,
+				Protocol:        "https",
+			},
+		},
+		HealthCheck: models.HealthCheck{
+			Target:             "HTTPS:443/path/to/site",
+			Interval:           60,
+			Timeout:            60,
+			HealthyThreshold:   3,
+			UnhealthyThreshold: 3,
+		},
+	}
+
+	readSGHelper(mockAWS, "l0-test-env_id-env", "env_sg")
+
+	createSGHelper(t, mockAWS, "l0-test-lb_id-lb", "vpc_id")
+	readSGHelper(mockAWS, "l0-test-lb_id-lb", "lb_sg")
+
+	for _, port := range req.Ports {
+		ingressInput := &ec2.AuthorizeSecurityGroupIngressInput{}
+		ingressInput.SetGroupId("lb_sg")
+		ingressInput.SetCidrIp("0.0.0.0/0")
+		ingressInput.SetIpProtocol("TCP")
+		ingressInput.SetFromPort(int64(port.HostPort))
+		ingressInput.SetToPort(int64(port.HostPort))
+
+		mockAWS.EC2.EXPECT().
+			AuthorizeSecurityGroupIngress(ingressInput).
+			Return(&ec2.AuthorizeSecurityGroupIngressOutput{}, nil)
+	}
+
+	iamRoleInput := &iam.CreateRoleInput{}
+	iamRoleInput.SetRoleName("l0-test-lb_id-lb")
+	iamRoleInput.SetAssumeRolePolicyDocument(provider.DEFAULT_ASSUME_ROLE_POLICY)
+
+	mockAWS.IAM.EXPECT().
+		CreateRole(iamRoleInput).
+		Return(&iam.CreateRoleOutput{}, nil)
+
+	renderedLBRolePolicy, err := provider.RenderLoadBalancerRolePolicy(
+		"region",
+		"123456789012",
+		"l0-test-lb_id",
+		provider.DEFAULT_LB_ROLE_POLICY_TEMPLATE)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	putIAMPolicyInput := &iam.PutRolePolicyInput{}
+	putIAMPolicyInput.SetPolicyName("l0-test-lb_id-lb")
+	putIAMPolicyInput.SetRoleName("l0-test-lb_id-lb")
+	putIAMPolicyInput.SetPolicyDocument(renderedLBRolePolicy)
+
+	mockAWS.IAM.EXPECT().
+		PutRolePolicy(putIAMPolicyInput)
+
+	listeners := make([]*elb.Listener, len(req.Ports))
+	for i, port := range req.Ports {
+		listener := &elb.Listener{}
+		listener.SetProtocol(port.Protocol)
+		listener.SetLoadBalancerPort(port.HostPort)
+		listener.SetInstancePort(port.ContainerPort)
+
+		if port.CertificateName != "" {
+			serverCertificateMetadataList := []*iam.ServerCertificateMetadata{
```

Comment:
[nit] this breaks the 
```go
input := ...
input.SetField(...)
```
pattern

Reply:
The purpose of this is to construct `listeners`, but in the process of constructing the array, `ListServerCertificates` is expected to be called if a certificate name is specified. `listeners` is used on [line 146](https://github.com/quintilesims/layer0/pull/410/files/a256730c903ef3983edfa569c55403e9284e50a1#diff-267fb5b55f4e244c24468db72e65e259R146) in the `SetListeners` method.

It looks strange, but it does follow that pattern, just in a delayed fashion due to `ListServerCertificates`.

Reply:
i mean it would be something like:
```
serverMeta := &iam.ServerCertMeta{}
serverMeta.SetArn(...)
serverMeta.SetServerCertificatName(...)

serverMetaList := []*iam.ServerMeta{serverMeta}
...
```